### ISSUE TITLE
Should we use a "default value" type class for Panic.panic?

### ISSUE THREAD

**Issue by sweirich** [antalsz/hs-to-coq#30]
Many uses of Panic.panic in CmmTypes.hs are with concrete result types (integer, width, etc).
 
If we have a class for default values, then we can safely parameterize the `panic` function with this class.
```
Parameter panic {a} `{Default a} : String -> a
```

**Comment by antalsz** [comment]
Is this actually the right design?  One thing we'll presumably want to prove is that `panic` is never called…

**Comment by sweirich** [comment]
The correctness theorem would prove that `panic` is never called.

I'm somewhat pessimistic that we will get this result just through our translation. It would take a lot of rewriting to replace all of the calls of panic with something else.  (And this is better than just returning the default value without wrapping it in panic.)

**Comment by antalsz** [comment]
Yeah, that's very fair.

**Comment by sweirich** [comment]
Ok, tried it out with my recent commit. 
commit 15c4060405095e4d75f6e18f23dd485ac3c5c3ed

**Comment by antalsz** [comment]
@sweirich How would that work with the "type class with no instances" approach?  I don't *think* that would damage the translation in 15c4060405095e4d75f6e18f23dd485ac3c5c3ed, and it might make the proofs easier.

**Comment by sweirich** [comment]
I don't understand your question. What do you mean the "type class with no instances" approach?

**Comment by sweirich** [comment]
BTW, there is probably a better place (and perhaps name) for the Default type class. Furthermore, we can tell Coq to be particularly aggressive about finding instances for it --- we will never care about coherence.

**Comment by antalsz** [comment]
Ah, sorry.  Something like this:

```coq
Class Default (a : Type) := { default : a }.

(* No instances  *)

Definition panic {a} `{Default : a} (_ : String) : a := default.
```

Of course, now I recall the problem with this, which is that we'd have to propagate the constraint to the types of the translated functions.  Never mind :upside_down_face:

**Comment by sweirich** [comment]
Yes, I was only intending this to be used with concrete types, not propagating instances. 
Furthermore, note that `panic` is a Parameter.  We never want to simplify it away during a proof because it marks a part of the program that is an error and we don't want to lose that information. But, we are guaranteed that this axiom is consistent because we could give panic a definition---precisely the one you have above.
Finally, if we extract, we can extract it back to the original panic.
